[게시판 회원 기능] 러너덕 미션 제출합니다

Deocksoo opened the pull request:
안녕하세요, 이번 미션 리뷰이 러너덕입니다. 
미션 제출이 늦어져 죄송합니다. 이번 미션 테스트 작성하기가 쉽지 않네요..
아직 테스트 코드에 중복도 많고 수정해야할 부분도 많은 것 같습니다 ㅜㅜ 
제출이 너무 늦어질 것 같아 일단 리뷰어님께 의견 받아보고 싶어서 부족한 채로 제출하게 되었습니다.
잘 부탁드립니다 :)

yk1028 committed:
feat: 실습

yk1028 committed:
feat: 환경설정

yk1028 committed:
feat: HTML 중복제거

yk1028 committed:
refactor: thymeleaf-layout 제거

yk1028 committed:
feat: static resource path 수정

yk1028 committed:
fix: devtools 개발 시에만 사용하도록 변경

yk1028 committed:
feat: 게시물 생성

yk1028 committed:
feat: 게시글 생성 후 전체 리스트 조회

yk1028 committed:
feat: 게시글 개별 조회

yk1028 committed:
style: formatting

yk1028 committed:
feat: Article에 id 추가

yk1028 committed:
feat: 게시글 수정 페이지로 이동

yk1028 committed:
feat: 게시글 수정

yk1028 committed:
feat: 게시글 삭제

yk1028 committed:
test: 테스트간 영향이 없도록 수정

yk1028 committed:
fix: UI 버그 수정

yk1028 committed:
style: formatting

yk1028 committed:
test: 테스트 코드 중복 제거

yk1028 committed:
[게시글 생성/조회/수정/삭제] 히브리 미션 제출합니다. (#16)

* feat: 실습

* feat: 환경설정

* feat: HTML 중복제거

* refactor: thymeleaf-layout 제거

* feat: static resource path 수정

* fix: devtools 개발 시에만 사용하도록 변경

* feat: 게시물 생성

* feat: 게시글 생성 후 전체 리스트 조회

* feat: 게시글 개별 조회

* style: formatting

* feat: Article에 id 추가

* feat: 게시글 수정 페이지로 이동

* feat: 게시글 수정

* feat: 게시글 삭제

* test: 테스트간 영향이 없도록 수정

* fix: UI 버그 수정

* style: formatting

* test: 테스트 코드 중복 제거

yk1028 committed:
test: ArticleRepositoryTests에서 서버를 띄우지 않도록 수정

yk1028 committed:
refactor: ArticleRepository 클래스의 clear 메서드 제거

yk1028 committed:
Merge remote-tracking branch 'woowacourse/yk1028' into yk1028

yk1028 committed:
refactor: Article id의 동시성 보장

yk1028 committed:
refactor: matchId 구현

yk1028 committed:
refactor: next id를 Article에서 관리하도록 수정

yk1028 committed:
test: Article test 추가

yk1028 committed:
refactor: @Autowired를 생성자 주입으로 변경

yk1028 committed:
refactor: JAP 활용 방식으로 변경

yk1028 committed:
feat: sql formatting 설정

yk1028 committed:
test: test 코드 수정

yk1028 committed:
docs: 기능목록 작성

Deocksoo committed:
Merge branch 'yk1028' of https://github.com/yk1028/jwp-blog into Deocksoo2

yk1028 committed:
refactor: matchId 메서드 삭제

yk1028 committed:
feat: DB를 H2에서 MySQL로 변경

yk1028 committed:
feat: Index 페이지에 대한 처리를 IndexController에서 담당하도록 수정

yk1028 committed:
feat: Login 페이지로 이동하는 기능 구현

UserController에서 유저와 관련된 control을 담당하도록 하였습니다.

yk1028 committed:
feat: 회원가입 페이지로 이동하는 기능 구현

yk1028 committed:
feat: 회원가입 기능 구현

Deocksoo committed:
Merge branch 'yk1028' of https://github.com/yk1028/jwp-blog into Deocksoo2

yk1028 committed:
feat: email 중복 검사

yk1028 committed:
feat: email 중복 시 errorMessage를 Model에 담아서 페이지에 노출

yk1028 committed:
feat: sign-up 페이지에서 좌측 상단 로고 누르면 홈으로 이동하도록 수정

yk1028 committed:
feat: 이름 규칙 적용

이름은 2~10자로 제한하며 숫자나 특수문자가 포함될 수 없다.

yk1028 committed:
feat: 비밀번호 규칙 적용

비밀번호는 8자 이상의 소문자, 대문자, 숫자, 특수문자의 조합이다.

yk1028 committed:
feat: 비밀번호 확인 기능 구현

yk1028 committed:
feat: 유저 목록 페이지 구현

yk1028 committed:
feat: FE에서 비밀번호 검증 구현

yk1028 committed:
feat: Login 기능을 User로부터 분리

yk1028 committed:
feat: http header/body logging

yk1028 committed:
feat: Login 기능 구현

yk1028 committed:
feat: 로그인 실패 시 상황에 맞는 실패 메시지를 띄운다.

- 이메일이 없는 경우
- 비밀번호가 틀린 경우

yk1028 committed:
feat: 로그아웃 기능 구현

yk1028 committed:
feat: 로그인 한 유저가 로그인/회원가입 화면에 접근할 경우 메인 화면을 띄운다.

yk1028 committed:
feat: 마이 페이지 구현

yk1028 committed:
feat: mypage-edit으로 이동하는 기능 구현

yk1028 committed:
feat: User 이름 수정 기능 구현

yk1028 committed:
feat: LogOut 상태에서 article을 생성할 수 없도록 수정

yk1028 committed:
feat: login 상태에서만 article 작성 가능하도록 수정

yk1028 committed:
feat: 자신의 Article만 수정 가능하도록 구현

yk1028 committed:
feat: 자신의 Article만 삭제 가능하도록 구현

yk1028 committed:
feat: id에 해당하는 User의 mypage를 보여주도록 수정

yk1028 committed:
feat: 자신의 회원 정보만 수정할 수 있도록 구현

yk1028 committed:
feat: 회원 탈퇴기능 구현

yk1028 committed:
refactor: Optional.get() 사용 부분 변경

yk1028 committed:
feat: User 이름 수정시 유효성 검사

Deocksoo committed:
Merge branch 'yk1028' of https://github.com/yk1028/jwp-blog into Deocksoo2

yk1028 committed:
refactor: User 이름 수정시 유효성 검사 예외처리 방법 수정

UserArgumentException class 생성

yk1028 committed:
refactor: UserRepository에서 조회 결과를 Optional로 반환하도록 수정

yk1028 committed:
test: User test 추가

- edit, delete에 대한 테스트를 추가하였습니다.
- h2 db로 변경 및 초기 데이터 추가

Deocksoo committed:
Merge branch 'yk1028' of https://github.com/yk1028/jwp-blog into Deocksoo2

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
전체적인 구현 👍 테스트에 정말 많은 공을 들이신것 같아요. 👍 
몇가지 피드백 추가했는데요,
참고 부탁드리고 궁금하신 점 있으시면 연락주세요!

Deocksoo committed:
fix: LoginService Test에서 userRepository를 모두 지워버리면서 발생한 버그를 수정합니다

after each 메서드에서 매번 모든 데이터를 지우고 다시 테스트하도록 만들었었는데,
이로 인해 다른 테스트에서 필요한 데이터까지 모두 지워버리게 되는 문제가 있었습니다.
이를 수정합니다.

Deocksoo committed:
refactor: LoginService의 logIn 메서드에서 불필요한 검증 로직을 제거합니다

Deocksoo committed:
refactor: logback-access.xml 파일의 위치를 resources로 변경합니다

Deocksoo committed:
feat: error.html 템플릿 파일을 추가합니다

예외 상황에서 보여줄 페이지를 지정하고 ControllerAdvice를 통해 요청을 처리해줄 예정입니다.

Deocksoo committed:
refactor: UserService에서 담당하던 User관련 validate를 UserFactory로 이동합니다

1. UserFactory를 만들고 도메인에서 User생성과 관련된 일들을 모두 담당하도록 하였습니다.
2. Duplicate 체크의 경우 repostiory를 확인해야 하므로 이동하지 않았습니다.
3. UserPublicInfoDto의 이름을 UserProfileDto로 변경하였습니다.

Deocksoo committed:
refactor: UserService에서 Article에 종속성을 갖지 않도록 수정합니다

UserService에서 Article에 의존하게 하지 않기 위해 ArticleService를 만들고,
UserController에서 ArticleService에 Article의 delete요청을 하도록 수정하였습니다.

Deocksoo committed:
[REFACTOR] Interceptor를 추가합니다

Login 여부를 Interceptor에서 확인한 뒤, controller에게 넘겨주도록 수정하였습니다.

Deocksoo committed:
[STYLE] build gradle 파일 의존성들의 순서를 정리합니다

Deocksoo committed:
[REFACTOR] Test하는 경우 사용할 DB와 배포하는 경우 사용할 DB를 각각 구분해줍니다

Deocksoo committed:
[REFACTOR] ControllerAdvice를 추가합니다

1. UserController와 ArticleController에서 발생하는 예외에 대한 처리를 ControllerAdvice에서
처리해주도록 수정합니다.
2. 예외 상황인 경우 에러 페이지를 보여줄 수 있도록 error.html 파일을 추가하였습니다.

Deocksoo committed:
[REFACTOR] isLoggedInUser 메서드에서 login 체크 로직을 Interceptor로 분리합니다

Deocksoo committed:
[REFACTOR] UserServiceTest에서 담당하던 validation 체크 테스트를 UserFactoryTest로 이동합니다

Deocksoo committed:
[REFACTOR] UserService에 findById 메서드를 만들고, 테스트를 추가합니다

1. save, update 메서드에서 동작을 수행한 후에 id를 리턴하도록 수정하였습니다.
2. 기존에 없었던 Service의 비즈니스 로직 테스트들을 추가하였습니다.

Deocksoo committed:
[REFACTOR] LoginController의 예외 처리 handler를 controllerAdvice로 이동합니다

Deocksoo committed:
[REFACTOR] RequestMapping 어노테이션을 활용해 url의 중복을 제거합니다

Deocksoo committed:
[REFACTOR] JSR-303 빈 검증 방식을 활용해 UserDto에서 User정보를 확인하도록 합니다

Deocksoo committed:
[REFACTOR] JSR-303 빈 검증에 필요한 의존성을 추가합니다

Deocksoo committed:
[STYLE] wihtespace를 convetion에 맞춰 포매팅합니다

Deocksoo committed:
[REFACTOR] Test시에 Login 후 jsessionId를 얻어오는 기능을 Utils 클래스가 담당하도록 합니다

Deocksoo committed:
[REFACTOR] MyPageController에서 중복되는 validation 코드를 제거합니다

login 되어있는 유저인지는 Interceptor에서, 수정하려는 User page에 접근하는 user가
valid한 user인지에 대한 체크는 UserController에 이미 만들어놓은 validateUser 메서드를
사용하도록 하였습니다.

Deocksoo committed:
[REFACTOR] HttpServeltRequest를 없애고, 인자에서 직접 HttpSession을 받아오도록 수정합니다

Deocksoo committed:
[REFACTOR] LogInControllerTest와 UserControllerTest에서 의미 없이 선언된 변수를 제거합니다

Deocksoo committed:
[DOCS] Todo List를 업데이트합니다

Deocksoo reviewed (commented):
안녕하세요 화투, 
피드백 남겨주신 부분들을 어떻게든 반영해보려고 하다가 제출이 너무 늦어지고 말았네요ㅜㅜ
아직도 부족한 부분이 너무 많은데, 시간이 이미 많이 지나버려서 일단 제출합니다.
저번 피드백이 기억나시지 않을 수도 있을것 같아서 크게 수정한 부분들을 한번 정리해 봤어요.
난리법석인 코드를 보시는 데에 조금이나마 도움이 되었으면 좋겠습니다..

1. Interceptor를 활용해 login 되어있는지 여부를 컨트롤러 앞단에서 체크하도록 수정했습니다.
2. UserService에서 validation 체크하는 부분을 모두 UserFactory를 만들어 옮기고, 비즈니스 로직과 관련된 기능만 수행하도록 수정하였습니다. 그리고 Dto에서 유효성을 검사하도록 하는 JSR-303 빈 검증 방식을 활용해서 체크를 한번 더 해 보았어요. 이렇게 두번씩 체크를 해도 괜찮을까요? 아니면 Dto에서 검증하는 경우에는 따로 domain에서 체크를 하지 않아도 괜찮은건가요?

3. 테스트에서 중복되는 login 코드들을 Utils 클래스로 뽑아내 보았습니다. 하지만 아직도 영 좋지 못한 냄새가 많이 나기는 해서 찝찝한 부분이 많아요. WebTestClient에 이어져있는 chaining도 어떻게 끊는 것이 좋을지 고민입니다. 차차 공부해서 더 수정해보고 싶어요.

피드백 남겨주신 부분들을 수정하면서 많은 공부가 된 것 같아요.
신경써 피드백 남겨주셔서 감사합니다!

Deocksoo committed:
[REFACTOR] ArticleControllerTest의 body가 data를 포함하고 있는지 직접 체크하도록 합니다

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
피드백 반영하시느라 정말 수고 많으셨습니다. :)
질문에 대한 답변 및 몇가지 피드백 추가했는데
다음 단계 진행하시면서 같이 참고 부탁드립니다. :)
